scapy/contrib/mqtt.py

```diff
@@ -40,119 +44,122 @@ def getfield(self, pkt, s):
                                       self.__class__.__name__)
 
 
-
 # LAYERS
-            
-control_packet_types = {1:'CONNECT (1)',
-                        2:'CONNACK (2)',
-                        3:'PUBLISH (3)',
-                        4:'PUBACK (4)',
-                        5:'PUBREC (5)',
-                        6:'PUBREL (6)',
-                        7:'PUBCOMP (7)',
-                        8:'SUBSCRIBE (8)',
-                        9:'SUBACK (9)',
-                        10:'UNSUBSCRIBE (10)',
-                        11:'UNSUBACK (11)',
-                        12:'PINGREQ (12)',
-                        13:'PINGRESP (13)',
-                        14:'DISCONNECT (14)'}
-
-
-QOS_level = {0:'0 (At most once delivery)',
-             1:'1 (At least once delivery)',
-             2:'2 (Exactly once delivery)'}
 
+control_packet_types = {1: 'CONNECT',
+                        2: 'CONNACK',
+                        3: 'PUBLISH',
+                        4: 'PUBACK',
+                        5: 'PUBREC',
+                        6: 'PUBREL',
+                        7: 'PUBCOMP',
+                        8: 'SUBSCRIBE',
+                        9: 'SUBACK',
+                        10: 'UNSUBSCRIBE',
+                        11: 'UNSUBACK',
+                        12: 'PINGREQ',
+                        13: 'PINGRESP',
+                        14: 'DISCONNECT'}
+
+
+QOS_level = {0: 'At most once delivery',
+             1: 'At least once delivery',
+             2: 'Exactly once delivery'}
 
+
+# source: http://stackoverflow.com/questions/43721828/scapy-referencing-multiple-fields-from-a-fieldlenfield/43722441#43722441
 class MQTT(Packet):
     name = "MQTT fixed header"
     fields_desc = [
         BitEnumField("type", 1, 4, control_packet_types),
-        BitEnumField("DUP", 0, 1, {0:'Disabled (0)',
-                                   1:'Enabled (1)'}),
+        BitEnumField("DUP", 0, 1, {0: 'Disabled',
+                                   1: 'Enabled'}),
         BitEnumField("QOS", 0, 2, QOS_level),
-        BitEnumField("RETAIN", 0, 1, {0:'Disabled (0)',
-                                      1:'Enabled (1)'}),
+        BitEnumField("RETAIN", 0, 1, {0: 'Disabled',
+                                      1: 'Enabled'}),
+        # Since the size of the len field depends on the next layer, we need to "cheat" with the length_of parameter
+        # and use adjust parameter to calculate the value.
```

Comment:
Same here, please link to the source: https://stackoverflow.com/a/43722441/3223422